modules/core/04-channel/keeper/keeper_test.go

```diff
@@ -541,3 +542,323 @@ func (suite *KeeperTestSuite) TestUnsetParams() {
+					suite.chainA.GetContext(),
+					path.EndpointA.ChannelConfig.PortID,
+					path.EndpointA.ChannelID,
+					5,
+				)
+				suite.Require().NoError(err)
+				// We expect 0 to be pruned and 0 left.
+				suite.Require().Equal(uint64(0), pruned)
+				suite.Require().Equal(uint64(0), left)
+
+				// we _do not_ expect error, simply a fast return
+				postPruneExpState(12, 12, 6)
+			},
+			nil,
+		},
+		{
+			"failure: packet sequence start not set",
+			func() {},
+			func() {
+				path.EndpointA.ChannelConfig.PortID = "portidone"
+			},
+			func(_, _ uint64) {},
+			types.ErrPruningSequenceStartNotFound,
+		},
+		{
+			"failure: packet sequence end not set",
+			func() {},
+			func() {
+				store := suite.chainA.GetContext().KVStore(suite.chainA.GetSimApp().GetKey(exported.StoreKey))
+				store.Delete(host.PruningSequenceEndKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID))
+			},
+			func(_, _ uint64) {},
+			types.ErrPruningSequenceEndNotFound,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			// Defaults will be filled in for rest.
+			upgradeFields = types.UpgradeFields{Version: ibcmock.UpgradeVersion}
+			limit = 10
+
+			// perform pre upgrade ops.
+			tc.pre()
+
+			suite.UpgradeChannel(path, upgradeFields)
+
+			tc.malleate()
+
+			pruned, left, err := suite.chainA.App.GetIBCKeeper().ChannelKeeper.PruneStalePacketState(
+				suite.chainA.GetContext(),
+				path.EndpointA.ChannelConfig.PortID,
+				path.EndpointA.ChannelID,
+				limit,
+			)
+
+			suite.Require().ErrorIs(err, tc.expError)
+
+			// check on post state.
+			tc.post(pruned, left)
+		})
+	}
+}
+
+// UpgradeChannel performs a channel upgrade given a specific set of upgrade fields.
+// Question(jim): setup.coordinator.UpgradeChannel() wen?
+func (suite *KeeperTestSuite) UpgradeChannel(path *ibctesting.Path, upgradeFields types.UpgradeFields) {
+	// configure the channel upgrade version on testing endpoints
+	path.EndpointA.ChannelConfig.ProposedUpgrade.Fields = upgradeFields
+	path.EndpointB.ChannelConfig.ProposedUpgrade.Fields = upgradeFields
+
+	err := path.EndpointA.ChanUpgradeInit()
+	suite.Require().NoError(err)
+
+	err = path.EndpointB.ChanUpgradeTry()
+	suite.Require().NoError(err)
+
+	err = path.EndpointA.ChanUpgradeAck()
+	suite.Require().NoError(err)
+
+	err = path.EndpointB.ChanUpgradeConfirm()
+	suite.Require().NoError(err)
+
+	err = path.EndpointA.ChanUpgradeOpen()
+	suite.Require().NoError(err)
+
+	err = path.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+}
+
+// sendMockPacket sends a packet from source to dest and acknowledges it on the source (completing the packet lifecycle).
+// Question(jim): find a nicer home for this?
+func (suite *KeeperTestSuite) sendMockPackets(source, dest *ibctesting.Endpoint, numPackets int) {
```

Comment:
we should be able to replace endponts with passing the path. and then use `path.RelayPacket`

will look into it with @chatton

Reply:
https://github.com/cosmos/ibc-go/pull/5444/commits/998e7ba5096121ea3927a8b0ad2a4b87c7f58070